qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/extensions/renarde/UriNamespaceResolverSupport.java

```diff
@@ -0,0 +1,131 @@
+/*******************************************************************************
+* Copyright (c) 2022 Red Hat Inc. and others.
+* All rights reserved. This program and the accompanying materials
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package com.redhat.qute.jdt.internal.extensions.renarde;
+
+import static com.redhat.qute.jdt.internal.QuteJavaConstants.JAVA_LANG_OBJECT_TYPE;
+import static com.redhat.qute.jdt.internal.extensions.renarde.RenardeJavaConstants.RENARDE_CONTROLLER_TYPE;
+
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.jdt.core.Flags;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.IType;
+import org.eclipse.jdt.core.ITypeHierarchy;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.core.search.SearchMatch;
+import org.eclipse.jdt.core.search.SearchPattern;
+
+import com.redhat.qute.commons.datamodel.resolvers.ValueResolverInfo;
+import com.redhat.qute.jdt.template.datamodel.AbstractDataModelProvider;
+import com.redhat.qute.jdt.template.datamodel.SearchContext;
+import com.redhat.qute.jdt.utils.JDTTypeUtils;
+
+/**
+ * uri, uriabs renarde support.
+ * 
+ * @author Angelo ZERR
+ * 
+ * @see https://github.com/quarkiverse/quarkus-renarde/blob/main/docs/modules/ROOT/pages/index.adoc#obtaining-a-uri-in-qute-views
+ *
+ */
+public class UriNamespaceResolverSupport extends AbstractDataModelProvider {
+
+	private static final Logger LOGGER = Logger.getLogger(UriNamespaceResolverSupport.class.getName());
+
+	private static final String URI_NAMESPACE = "uri";
+
+	private static final String URIABS_NAMESPACE = "uriabs";
+
+	@Override
+	public void beginSearch(SearchContext context, IProgressMonitor monitor) {
+		IJavaProject javaProject = context.getJavaProject();
+		IType type = JDTTypeUtils.findType(javaProject, RENARDE_CONTROLLER_TYPE);
+		if (type != null) {
+			try {
+				// Find all classes which extends 'io.quarkiverse.renarde.Controller'
+				ITypeHierarchy typeHierarchy = type.newTypeHierarchy(monitor);
+				IType[] controllerTypes = typeHierarchy.getAllClasses();

```

Comment:
@rgrunber here I need to collect all classes which extends "io.quarkiverse.renarde.Controller" class. I have a usecase with : 

 * Application which extends Controller
 * Todos which extends ControllerWithUser which extends Controller.

When Controller and ControllerWithUser comes from sources (I copied/pasted thoses sources in my project), I can get Application and Todos.

But when Controller and ControllerWithUser comes from JAR,it seems (according the problem with @fbricon with Todos), that it get only Application (notTodos which extends ControllerWithUser which extends Controller).

At first do you think my code is correct? Have you some idea about this problem?

Reply:
Is it the same result if you try using `IType[] getAllSubtypes(IType type);` ?

Reply:
The first thing I notice from running the test case is that 

```
/qute-renarde/lib/quarkus-renarde-1.0.0-SNAPSHOT.jar[CPE_LIBRARY][K_BINARY][isExported:false][attributes:module=true]
/qute-renarde/lib/quarkus-renarde-oidc-1.0.0-SNAPSHOT.jar[CPE_LIBRARY][K_BINARY][isExported:false][attributes:module=true]
```

don't exist. They're defined in the `.classpath` so they show, but there's no `lib` folder in the qute-renard project so no way to even get the type.